### ISSUE TITLE
Error uploading AIML

### ISSUE THREAD

**Issue by geezus2016** [Program-O/Program-O#186]
I have installed Program-O 2.6.3 successfully.

I am now at the stage of uploading aiml. I downloaded the aiml-en-us-foundation-alice.v1-6-master files.

When I upload them, I am getting errors:

xxxxxx changed below.

Error 1871: Element '{http://www.alicebot.org/TR/2001/WD-aiml}category': This element is not expected. in /home/xxxxxx/public_html/_chatbot/admin/
on line 40

Error 1839: Element '{http://www.alicebot.org/TR/2001/WD-aiml}a', attribute 'target': [facet 'pattern'] The value '_new' is not accepted by the pattern '_(blank|self|parent|top)|[A-Za-z]\c*'. in /home/xxxxxxx/public_html/_chatbot/admin/
on line 162

Error 1824: Element '{http://www.alicebot.org/TR/2001/WD-aiml}a', attribute 'target': '_new' is not a valid value of the atomic type '{http://www.alicebot.org/TR/2001/WD-aiml}FrameTarget'. in /home/xxxxxx/public_html/_chatbot/admin/
on line 162

I have no idea what this means but I cannot upload any aiml files, as I get the above error messages (more than I have shown) with each upload.

Any ideas?

**Comment by Dave-Morton** [comment]
Several questions present themselves here:

1.) Can you please provide your system specs? [this wiki entry](https://github.com/Program-O/Program-O/wiki/Reporting-Issues) can guie you through the necessary steps

2.) Did you download your copy of Program O from the MASTER branch, or the DEV branch? It makes a difference at this stage of the game.

3.) Can you please tell us where you downloaded the AIML files from? They're available from many sources, and some are better than others.

Once we have this information I can better advise you as to what's going on. Thanks.

**Comment by geezus2016** [comment]
Hi Dave,

Hope this helps:

PC:
Windows 10
Chrome browser

Web:
Apache Version	2.2.31
PHP Version	5.4.45
MySQL Version	5.5.51-38.2

Downloads:
Master was selected when I downloaded program-o 2.6.3.

aimls from https://github.com/Urthen/aiml-en-us-foundation-alice.v1-6?files=1

Fixes:
I have found a lot of the aimls can be fixed with changing "_new" to "blank"

Previous:
I had an older version of program-o installed (2.4.7) and the aiml files I had then uploaded fine.

I have since removed 2.4.7 and the database some time ago.  Yesterday I decided to start again from scratch with the updated 2.6.3 version.

As stated above, all went well, installed just fine. Filled in all the bot personality etc.

Then I tried to upload my original aim files from my 2.4.7 install. They failed!

So I looked for other aimls and found the ones at the site listed above. But alas, same errors.

Just curious if it is something I am doing wrong this time round as I did not have any issues with 2.4.7 in the past.

Thanks for any input Dave :)

**Comment by Dave-Morton** [comment]
The current version of the MASTER branch, while stable, is missing some features that the DEV branch has, namely using a different DTD and XML Schema file that have been modified to accept HTML tags. The DEV branch also has the option of skipping AIML validation for uploaded files (though skipping validation isn't really recommended, due to the possibility of invalid AIML causing problems with your chatbot). If an AIML file is failing validation because of included HTML code, then skipping validation is harmless.

Bottom line here is that you should give the DEV branch a try and see if that takes care of the problem. I've been VERY busy with school, so haven't had the time to work on merging branches, but the end of the semester is only a few weeks away and I should have more time then to get the MASTER branch up to speed. :)

**Comment by geezus2016** [comment]
Thanks Dave for the reply.

Quick question, how do I install the Dev version over the master? Or do I need to do a clean install?

Many thanks.

**Comment by geezus2016** [comment]
Running into issues with Dev version!

The config folder cannot be written to, or does not exist. Please correct this before you continue.
The debug folder cannot be written to, or does not exist. Please correct this before you continue.
The logs folder cannot be written to, or does not exist. Please correct this before you continue.

![program-o](https://cloud.githubusercontent.com/assets/24223311/20703748/ca7db770-b614-11e6-8a10-1046da4b2be1.png)

Permissions for the above folders are all as shown in the above image. Yet I still see those messages.

I have tried and tried to reinstall the dev version. No matter what, those errors are always showing.
(Clean install in another folder from the Master install. Master installed just fine, no folder permission issues as above, so I am unsure why the dev version just will not install)

**Comment by Dave-Morton** [comment]
That's a new one for me. I assume you're using a WAMP package, given it's a Windows server? My only suggestion would be to check the box at the top that says "Apply changes recursively to subfolders and files", but frankly, all of those folders should already exist, and Windows systems generally don't have the permissions problems that Linux systems do. I've run through a complete, clean install of Program O from the master branch, to see if I can replicate the problems you're experiencing with the uploading of the AIML files, and I've noticed that the vast majority of the files in the zip file you've linked to _DO_, indeed, upload just fine. The bugs in the upload script within the MASTER branch version are a little misleading, and fail to report the filenames of the 9 files that actually failed validation, but the remaining 50 files not only passed validation, but were successfully uploaded and added to the database. I'm now going to attempt an overlay install of Program O from the DEV branch, to see if I get the same permissions errors, so that I can best advise you on how to proceed. Gimme a couple, please. :)

**Comment by Dave-Morton** [comment]
Ok, I've found the issue with those messages being displayed, and that bug will be fixed shortly, but in your case it's not really a huge issue at present. a simple "overlay install" is all that's required, which means that all that's necessary to upgrade is to copy the files from the DEV branch directly over the existing install from the MASTER branch and proceed as normal. You don't have to run the install script once the files are copied.

As for the bug that's causing those messages to be displayed, there's a typo in the file `install_config.php` that's causing the issue, and as I mentioned, I'll be correcting that shortly. There are some unfinished code fixes and features that I have to sort out before I can update the DEV branch, and I don't want to have to revert dozens of files just to correct a single typo. I'll try to get those unfinished fixes and features completed today, and maybe push the changes (including the typo fix) by this evening or tomorrow. Meanwhile, I have a suggestion for you: Try using a different AIML set. I have a couple of them that _should_ pass validation on my website at http://www.geekcavecreations.com/?downloads - Look for `Unicode_ALICE_AIML.zip` or just `aiml.zip`. either of those should work for you flawlessly.  Please let me know how you make out. :)

**Comment by geezus2016** [comment]
Hi Dave. I am not running my bot from windows. It's online. The screengrab was from FlashFXP on changing the attributes. Trust me, i have checked the "apply to sub folders..." also. I managed to get the DEV version to work, by simply copying my global_config.php from the MASTER install to the DEV's/config folder. I have also been able to edit aiml's and upload and the dev version is working well.

I am unsure why it kept saying those folders cannot be written to, or do not exist. When clearly they are just fine :)

Anyway, it would be nice to know about the error aiml files that were causing problems for the MASTER. As I said, most errors were caused by "_new" and simply changing to "blank" fixed those. I also noticed some spelling errors (I think and assume) such as <peron> instead of <person> in a few.

Again, easy to fix and that was all thanks to the great Program O Validator which helped track down the lines with errors.

Just a quick question again, in case of updated versions/releases, how do I install Program-O over the one I have up and running? I only see instructions for a clean install included.

Thanks again Dave for all you do. Amazing job as always.

hi5

**Comment by Dave-Morton** [comment]
the problem with that was a typo in the file `install_config.php` where `_BASE_PATH_` was coded in as `_BASE_DIR_` instead. That was causing the script to check the wrong folder path, and of course since the wrong path was being used, the folders in that wrong path didn't exist, so they were being incorrectly reported as unwritable. The most recent release fixed the typos so should now work.

However, as I mentioned previously, you should not have to re-run the install script in order to take advantage of the new features and fixes. A simple copying of the files over to your current (previously working - mostly) install should fix the bulk of the upload problem. It won't fix the invalid AIML files, of course, but the error messages about those invalid AIML files will be a __LOT__ easier to read and understand.

**Comment by geezus2016** [comment]
Glad you fixed the Path/Dir bug :) At least I wasn't doing something wrong, like I usually do lol

Surprised nobody else mentioned it.

I somehow missed your previous message about the fix and aiml set. Apologies for asking about how to install an update twice! DOH!

I shall of course try the aiml set(s) you provided the link for. However, right now, both Chrome and IE refuse to download them. Right clicking "save as" only attempts to download html downloadme. Left clicking on them opens up a page. Happens in both browsers. I shall try again later.

Keep up the great work Dave.

Feel free to close this?

**Comment by Dave-Morton** [comment]
It looks as if my download script has blown up. I'll check into the reason for it. Meanwhile, here's a direct link to one of the files:

http://www.geekcavecreations.com/Downloads/aiml.zip

I don't want to close this issue until we're both satisfied that the issue has been fully addressed, so please let me know if you have any further trouble with this. Thanks.

**Comment by geezus2016** [comment]
Hi Dave,

Thanks for all your support. I am more than happy. I have the DEV version up and running and successfully uploaded aiml files to it etc. You have fixed the install issue with Dev and I am more than happy here :)

Thanks for the zip file above too. Glad I could bring another little problem your way with downloads on your site !! All I seem to do is find little faults lol Sorry.

Feel free to close this if you are happy too.

Again, thank you.

**Comment by Dave-Morton** [comment]
You're most welcome. If you run across anything else, just let me know.
